### PR TITLE
Asb performance

### PR DESCRIPTION
**Pull request by danielmarbach**
Closes https://github.com/Particular/docs.particular.net/issues/2939

@Particular/azure-maintainers thoughts?

### CODE REVIEW

File: Snippets/ASB/ASB_7/Performance.cs [hunk 1]
```diff
@@ -23,7 +23,7 @@ await endpointInstance.Send(new SomeMessage())
         #endregion
 
         #region asb-fast-send
-        var tasks = new List<Task>();
+        var tasks = new List<Task>(10000);
```
Comment (SeanFeldman) [inline Snippets/ASB/ASB_7/Performance.cs]: Would it make sense to extract the value from here and loop and call it with something to disclose the intention (number-of-messages-to-send)?

File: Snippets/ASB/ASB_8/Performance.cs [hunk 2]
```diff
@@ -23,7 +23,7 @@ await endpointInstance.Send(new SomeMessage())
         #endregion
 
         #region asb-fast-send
-        var tasks = new List<Task>();
+        var tasks = new List<Task>(10000);
```
Comment (SeanFeldman) [inline Snippets/ASB/ASB_8/Performance.cs]: Same as above

File: transports/azure-service-bus/performance-tuning_body_asb_[7,).partial.md [hunk 3]
```diff
@@ -22,39 +22,40 @@ Snippet: asb-fast-send
...
-  * `NumberOfClientsPerEntity()`: Keep this value equal to the number of factories, this ensures there is 1 internal receive client per factory.
+  * `MessagingFactories().NumberOfMessagingFactoriesPerNamespace()`: Each factory manages a TCP connection to the broker, this connection has throughput limits. The number of factories depends on how much bandwidth Azure Service Bus is allowed to consume and the size of the messages. Monitor the networks received and sent bytes per second. If the network isn't congested, try to increase the number of factories and clients. The optimal value range is probably somewhere between 16 and a multiple of it, like 64 on partitioned queues (as they consist of 16 partitions).
```
Comment (SeanFeldman) [inline transports/azure-service-bus/performance-tuning_body_asb_[7,).partial.md]: >  The number of factories depends on how much bandwidth Azure Service Bus is allowed to consume and the size of the messages.

Now reading this is really confusing. Factories control how much bandwidth the  ASB **client** will have when communicating to the broker.

File: transports/azure-service-bus/performance-tuning_body_asb_[7,).partial.md [hunk 4]
```diff
@@ -22,39 +22,40 @@ Snippet: asb-fast-send
...
   * `MessageReceivers().PrefetchCount()`: Value is set per receiver. The higher throughput rate for an individual receiver, the more aggressive prefetching should be applied (i.e. the higher value for `PrefetchCount`).
-  * `Transactions()`: The transactional guarantees have a significant impact on performance, for example  the application using `ReceiveOnly` will generally have much better performance than application using SendsAtomicWithReceive`.
+  * `Transactions()`: The transactional guarantees have a significant impact on performance, for example,  the application using `ReceiveOnly` will have much better performance than an application using SendsAtomicWithReceive`.
```
Comment (SeanFeldman) [inline transports/azure-service-bus/performance-tuning_body_asb_[7,).partial.md]: > impact on performance, for example

"...impact on performance **. F**or example"

File: transports/azure-service-bus/performance-tuning_body_asb_[7,).partial.md [hunk 5]
```diff
@@ -22,39 +22,40 @@ Snippet: asb-fast-send
...
 
+NOTE: If an endpoint is a sender and a receiver some settings like `MessagingFactories().NumberOfMessagingFactoriesPerNamespace()` need to be tweaked accordingly. In general it is recommended to optimize for receive performance since these settings might also improve the send performance.
```
Comment (SeanFeldman) [inline transports/azure-service-bus/performance-tuning_body_asb_[7,).partial.md]: > some settings like

"the following settings"

File: transports/azure-service-bus/performance-tuning_body_asb_[7,).partial.md [hunk 6]
```diff
@@ -22,39 +22,40 @@ Snippet: asb-fast-send
...
+NOTE: When send operations occur inside a handler the default behavior is to add the operation to a batch, which will be executed after the handler function returns. Unless the send options request for immediate dispatch, in such a case the execution is immediate. Therefore using an `await` on the `Send()` operation has a negative impact when used on an immediate dispatch, but has no impact on the default settings. It is advised to always return the task instead of awaiting the task never to be impacted by this difference in behavior, which might lead to subtle issues over time. To learn more about batched dispatch refer to the [Batched message dis](/nservicebus/messaging/batched-dispatch.md) article.
```
Comment (SeanFeldman) [inline transports/azure-service-bus/performance-tuning_body_asb_[7,).partial.md]: > [Batched message dis]

looks like a truncated "disposition" word 🙂

File: transports/azure-service-bus/performance-tuning_body_asb_[7,).partial.md [hunk 7]
```diff
@@ -22,39 +22,40 @@ Snippet: asb-fast-send
...
 
   * Concurrency: The optimal values are in the range of 128 per client.
   * PrefetchCount: The optimal values are 1x or 2x the per receiver concurrency, so 128 or 256.
 
 The optimal numbers for Concurrency and PrefetchCount for receivers are impacted significantly by sends. If the average handler sends roughly the same compound size of messages that it receives, then the numbers recommended above will be optimal. 
 
-However, if the handler sends more messages or sends larger messages (e.g. in the range of 10 to 100 times the total data size per single receive), then Concurrency and PrefetchCount values should be lower in order to reduce incoming bandwidth and give more bandwidth to the senders. In such scenario:
+However, if the handler sends more messages or sends larger messages (e.g. in the range of 10 to 100 times the total data size per single receive), then Concurrency and PrefetchCount values should be lower to reduce incoming bandwidth and give more bandwidth to the senders. In such scenario:
```
Comment (SeanFeldman) [inline transports/azure-service-bus/performance-tuning_body_asb_[7,).partial.md]: >  then Concurrency and PrefetchCount values

single back tick around `PrefetchCount `

File: transports/azure-service-bus/performance-tuning_body_asb_[7,).partial.md (hunk 3 above)
Comment (danielmarbach) [inline transports/azure-service-bus/performance-tuning_body_asb_[7,).partial.md]: I think it should be slightly reworded to

The number of factories needed/required depends on..

**Comment by danielmarbach** [comment]
@SeanFeldman done

**Pull request merged**